src/internal/clusterstate/v2.10.0/pipelines.go

```diff
@@ -0,0 +1,239 @@
+		changes, ok := pipVersionChanges[pi.Pipeline.Name]
+		if !ok {
+			changes = make(map[uint64]uint64)
+			pipVersionChanges[pi.Pipeline.Name] = changes
+		}
+		changes[currVersion] = correctVersion
+	}
+	return updates, pipVersionChanges, nil
+}
+
+func UpdateJobPipelineVersions(ctx context.Context, tx *pachsql.Tx, pipVersionChanges map[string]map[uint64]uint64) error {
+	jobUpdates, err := collectJobUpdates(ctx, tx, pipVersionChanges)
+	if err != nil {
+		return err
+	}
+	if err := updateJobRows(ctx, tx, jobUpdates); err != nil {
+		return err
+	}
+	return nil
+}
+
+func collectJobUpdates(ctx context.Context, tx *pachsql.Tx, pipVersionChanges map[string]map[uint64]uint64) ([]*jobRow, error) {
+	rr, err := tx.QueryxContext(ctx, listJobInfos)
+	if err != nil {
+		return nil, errors.Wrap(err, "list jobs")
+	}
+	defer rr.Close()
+	ji := &pps.JobInfo{}
+	var updates []*jobRow
+	for rr.Next() {
+		var row jobRow
+		if err := rr.Err(); err != nil {
+			return nil, errors.Wrap(err, "row error")
+		}
+		if err := rr.StructScan(&row); err != nil {
+			return nil, errors.Wrap(err, "scan job row")
+		}
+		if err := proto.Unmarshal(row.Proto, ji); err != nil {
+			return nil, errors.Wrapf(err, "unmarshal proto")
+		}
+		if changes, ok := pipVersionChanges[ji.Job.Pipeline.String()]; ok {
+			if new, ok := changes[ji.PipelineVersion]; ok {
+				ji.PipelineVersion = new
+				data, err := proto.Marshal(ji)
+				if err != nil {
+					return nil, errors.Wrapf(err, "marshal job info %v", ji)
+				}
+				updates = append(updates, &jobRow{Key: row.Key, Proto: data})
+			}
+		}
+	}
+	return updates, nil
+}
+
+func UpdatePipelineRows(ctx context.Context, tx *pachsql.Tx, pipUpdates []*PipUpdateRow) error {
+	if len(pipUpdates) != 0 {
+		var pipValues string
+		for _, u := range pipUpdates {
+			pipValues += fmt.Sprintf(" ('%s', '%s', decode('%v', 'hex')),", u.Key, u.IdxVersion, hex.EncodeToString(u.Proto))
+		}
+		pipValues = pipValues[:len(pipValues)-1]
+		stmt := fmt.Sprintf(`
+                 UPDATE collections.pipelines AS p SET
+                   idx_version = v.idx_version,
+                   proto = v.proto
+                 FROM (VALUES%s) AS v(key, idx_version, proto)
+                 WHERE p.key = v.key;`, pipValues)
+		log.Info(ctx, "deduplicate pipeline versions statement", zap.String("stmt", stmt))
+		if _, err := tx.ExecContext(ctx, stmt); err != nil {
+			return errors.Wrapf(err, "update pipeline rows statement: %v", stmt)
+		}
+	}
+	return nil
+}
+
+func updateJobRows(ctx context.Context, tx *pachsql.Tx, jobUpdates []*jobRow) error {
+	if len(jobUpdates) != 0 {
+		var jobValues string
+		for _, u := range jobUpdates {
+			jobValues += fmt.Sprintf(" ('%s', decode('%v', 'hex')),", u.Key, hex.EncodeToString(u.Proto))
+		}
+		jobValues = jobValues[:len(jobValues)-1]
+		stmt := fmt.Sprintf(`
+                 UPDATE collections.jobs AS j SET
+                   proto = v.proto
+                 FROM (VALUES%s) AS v(key, proto)
```

Comment:
Does this get pretty big if there are a lot of jobs?  What's the most jobs that can be updated in one statement?  Is there a limit?

Reply:
this could potentially get big ~number of jobs. There's no specific limit in Postgres. Maybe I should batch these into 100 jobs at a time.